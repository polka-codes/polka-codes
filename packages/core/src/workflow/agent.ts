// generated by polka.codes

import type { LanguageModelV2 } from '@ai-sdk/provider'
import type { ModelMessage, UserContent } from 'ai'
import type { z } from 'zod'
import { toJSONSchema } from 'zod/v4'
import {
  AgentBase,
  type AgentNameType,
  type AgentPolicy,
  AnalyzerAgent,
  ArchitectAgent,
  CodeFixerAgent,
  CoderAgent,
  type ExitReason,
  type SharedAgentOptions,
} from '../Agent'
import type { ToolFormat } from '../config'
import { type FullToolInfoV2, type ToolResponse, type ToolResponseInterrupted, ToolResponseType } from '../tool'
import { UsageMeter } from '../UsageMeter'
import type { BaseStepSpec, Json, ParseOutputResult, StepRunResult, StepSpecHandler, TemplatedString, WorkflowContext } from './types'
import { resolveTemplatedString } from './utils'

export type AgentStepSpec<
  TInput extends Record<string, Json> = Record<string, Json>,
  TOutput extends Record<string, Json> = Record<string, Json>,
> = BaseStepSpec<TInput, TOutput> & {
  type: 'agent'
  messages: (UserContent | TemplatedString<TInput>)[]
  outputSchema?: z.ZodType<TOutput>
  parseOutput?: (raw: string) => ParseOutputResult<TOutput>
  provider?: string
  model?: string
  modelParameters?: Record<string, any>
  toolFormat?: ToolFormat
  agent?: AgentNameType
  tools?: FullToolInfoV2[]
  systemPrompt?: string | TemplatedString<TInput>
}

const agentRegistry: Record<string, new (options: SharedAgentOptions) => AgentBase> = {
  analyzer: AnalyzerAgent,
  architect: ArchitectAgent,
  coder: CoderAgent,
  codefixer: CodeFixerAgent,
}

class WorkflowAgent extends AgentBase {
  protected onBeforeInvokeTool(_name: string, _args: Record<string, string>): Promise<ToolResponse | undefined> {
    return Promise.resolve(undefined)
  }
}

export type AgentContextParameters = {
  toolFormat?: ToolFormat
  os?: string
  policies?: AgentPolicy[]
  modelParameters?: Record<string, any>
  scripts?: Record<string, string | { command: string; description: string }>
  retryCount?: number
  requestTimeoutSeconds?: number
  usageMeter?: UsageMeter
}

type ResumeState = {
  messages?: ModelMessage[]
  usage?: {
    input: number
    output: number
    cachedRead: number
    cost: number
    messageCount: number
  }
}

export const makeAgentStepSpecHandler = (
  getModelFn: (step: AgentStepSpec, context: WorkflowContext) => Promise<LanguageModelV2>,
): StepSpecHandler => ({
  type: 'agent',
  handler(step: AgentStepSpec) {
    return {
      ...step,
      async run(
        input: Record<string, Json>,
        context: WorkflowContext,
        resumedState?: ResumeState,
      ): Promise<StepRunResult<Record<string, Json>>> {
        const logger = context.logger ?? console
        if (context.verbose && context.verbose >= 1) {
          logger.log(`[agent-step] Running agent step '${step.id}' with input:`, input)
        }
        try {
          const model = await getModelFn(step, context)

          const parameters: AgentContextParameters = context.parameters ?? {}

          const usageMeter = parameters?.usageMeter ?? new UsageMeter()
          const toolFormat = step.toolFormat ?? parameters.toolFormat ?? 'native'
          const policies = parameters.policies ?? []
          const modelParameters = { ...(parameters.modelParameters ?? {}), ...(step.modelParameters ?? {}) }

          if (resumedState?.usage) {
            usageMeter.setUsage(resumedState.usage)
          }

          const getAgent = () => {
            if (step.agent) {
              const agentName = step.agent
              const AgentClass = agentRegistry[agentName]
              if (!AgentClass) {
                throw new Error(`Unknown agent: ${agentName}`)
              }

              if (context.verbose && context.verbose >= 1) {
                logger.log(`[agent-step] Using agent: ${agentName}`)
              }

              return new AgentClass({
                ai: model,
                os: parameters.os ?? 'linux',
                provider: context.provider,
                toolFormat,
                additionalTools: step.tools,
                policies,
                usageMeter,
                parameters: modelParameters,
                scripts: parameters.scripts,
                callback: context.agentCallback,
                requireToolUse: false,
                retryCount: parameters.retryCount,
                requestTimeoutSeconds: parameters.requestTimeoutSeconds,
              })
            } else {
              if (!step.systemPrompt) {
                throw new Error('No system prompt specified for the agent step.')
              }
              if (context.verbose && context.verbose >= 1) {
                logger.log(`[agent-step] Using generic WorkflowAgent`)
              }
              const systemPrompt = resolveTemplatedString(step.systemPrompt, input)
              return new WorkflowAgent('agent', model, {
                systemPrompt,
                tools: step.tools ?? [],
                toolNamePrefix: toolFormat === 'native' ? '' : 'tool_',
                provider: context.provider,
                agents: [],
                scripts: parameters.scripts,
                callback: undefined,
                policies,
                toolFormat,
                parameters: modelParameters,
                usageMeter,
                requireToolUse: false,
                retryCount: parameters.retryCount,
                requestTimeoutSeconds: parameters.requestTimeoutSeconds,
              })
            }
          }

          const agent = getAgent()

          if (resumedState?.messages) {
            agent.setMessages(resumedState.messages)
          }

          const userContentArray: UserContent[] = step.messages.map((message) => {
            if (typeof message === 'string' || Array.isArray(message)) {
              return message
            }
            return resolveTemplatedString(message, input)
          })

          const combinedContentParts: Exclude<UserContent, string> = []
          for (const content of userContentArray) {
            if (typeof content === 'string') {
              combinedContentParts.push({ type: 'text', text: content })
            } else {
              combinedContentParts.push(...content)
            }
          }

          if (context.verbose && context.verbose >= 1) {
            logger.log(`[agent-step] Starting agent with content:`, JSON.stringify(combinedContentParts, null, 2))
          }

          const exitReason = await agent.start(combinedContentParts)

          if (context.verbose && context.verbose >= 1) {
            logger.log(`[agent-step] Agent exited with reason:`, exitReason)
          }

          const handleExitReason = async (reason: ExitReason): Promise<StepRunResult<Record<string, Json>>> => {
            switch (reason.type) {
              case 'Pause':
                return {
                  type: 'paused',
                  state: {
                    messages: agent.messages,
                    usage: usageMeter.usage,
                  },
                }
              case 'UsageExceeded':
                return { type: 'error', error: new Error('Usage limit exceeded') }
              case ToolResponseType.Exit: {
                const raw = reason.message
                if (step.parseOutput) {
                  const result = step.parseOutput(raw)
                  if (result.success) {
                    if (step.outputSchema) {
                      const validationResult = step.outputSchema.safeParse(result.data)
                      if (validationResult.success) {
                        return { type: 'success', output: validationResult.data }
                      }
                      let errorMessage = `Output validation failed: ${validationResult.error.toString()}`
                      if (step.outputSchema) {
                        errorMessage += `\n\nExpected JSON schema:\n${JSON.stringify(toJSONSchema(step.outputSchema), null, 2)}`
                      }
                      const newReason = await agent.continueTask(errorMessage)
                      return handleExitReason(newReason)
                    }
                    return { type: 'success', output: result.data as any }
                  }
                  let errorMessage = result.error ?? 'Invalid format.'
                  if (step.outputSchema) {
                    errorMessage += `\n\nExpected JSON schema:\n${JSON.stringify(toJSONSchema(step.outputSchema), null, 2)}`
                  }
                  const newReason = await agent.continueTask(errorMessage)
                  return handleExitReason(newReason)
                } else if (step.outputSchema) {
                  try {
                    const output = JSON.parse(raw)
                    const validationResult = step.outputSchema.safeParse(output)
                    if (validationResult.success) {
                      return { type: 'success', output: validationResult.data }
                    }
                    let errorMessage = `Output validation failed: ${validationResult.error.toString()}`
                    errorMessage += `\n\nExpected JSON schema:\n${JSON.stringify(toJSONSchema(step.outputSchema), null, 2)}`
                    const newReason = await agent.continueTask(errorMessage)
                    return handleExitReason(newReason)
                  } catch (e) {
                    const error = e instanceof Error ? e : new Error(String(e))
                    let errorMessage = `Failed to parse agent output as JSON: ${error.message}`
                    errorMessage += `\n\nExpected JSON schema:\n${JSON.stringify(toJSONSchema(step.outputSchema), null, 2)}`
                    const newReason = await agent.continueTask(errorMessage)
                    return handleExitReason(newReason)
                  }
                }
                return { type: 'success', output: reason }
              }
              case ToolResponseType.HandOver:
              case ToolResponseType.Delegate:
                return { type: 'success', output: reason }
              case 'Aborted':
              case 'WaitForUserInput':
                return { type: 'error', error: new Error(`Agent exited for reason: ${reason.type}`) }
              case ToolResponseType.Interrupted:
                return { type: 'error', error: new Error(JSON.stringify((reason as ToolResponseInterrupted).message)) }
              default:
                // This should be unreachable if all cases are handled.
                return {
                  type: 'error',
                  error: new Error(`Agent exited for an unhandled reason: ${JSON.stringify(reason)}`),
                }
            }
          }

          return await handleExitReason(exitReason)
        } catch (e) {
          return { type: 'error', error: e instanceof Error ? e : new Error(String(e)) }
        }
      },
    }
  },
})
