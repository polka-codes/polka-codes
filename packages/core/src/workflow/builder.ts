import type { AgentStepSpec } from './agent'
import type { BaseStepSpec, BranchStepSpec, CustomStepSpec, Json } from './types'

class StepsBuilder<
  TInput extends Record<string, Json> = Record<string, Json>,
  TOutput extends Record<string, Json> = Record<string, Json>,
> {
  readonly #steps: BaseStepSpec[] = []

  public build(): BaseStepSpec<TInput, TOutput> {
    if (this.#steps.length === 0) {
      throw new Error('A workflow must have at least one step.')
    }

    const rootStep =
      this.#steps.length > 1
        ? {
            id: 'root',
            type: 'sequential',
            steps: this.#steps,
          }
        : this.#steps[0]

    return rootStep as BaseStepSpec<TInput, TOutput>
  }

  public step<TStepOutput extends Record<string, Json>, TStepSpec extends BaseStepSpec<TOutput, TStepOutput>>(
    step: TStepSpec,
  ): StepsBuilder<TInput, TStepOutput> {
    this.#steps.push(step)
    return this as unknown as StepsBuilder<TInput, TStepOutput>
  }

  public parallel<TStepOutput extends Record<string, Json>>(
    id: string,
    step: BaseStepSpec<TOutput, TStepOutput>,
  ): StepsBuilder<TInput, TStepOutput> {
    return this.step({
      id,
      type: 'parallel',
      step,
    })
  }

  public custom<TStepOutput extends Record<string, Json>>(spec: CustomStepSpec<TOutput, TStepOutput>): StepsBuilder<TInput, TStepOutput>
  public custom<TStepOutput extends Record<string, Json>>(
    id: string,
    run: CustomStepSpec<TOutput, TStepOutput>['run'],
  ): StepsBuilder<TInput, TStepOutput>
  public custom<TStepOutput extends Record<string, Json>>(
    idOrSpec: string | CustomStepSpec<TOutput, TStepOutput>,
    run?: CustomStepSpec<TOutput, TStepOutput>['run'],
  ): StepsBuilder<TInput, TStepOutput> {
    if (typeof idOrSpec === 'string') {
      if (!run) {
        throw new Error('Custom step "run" function must be provided.')
      }
      return this.step({
        id: idOrSpec,
        type: 'custom',
        run,
      })
    }
    return this.step(idOrSpec)
  }

  public agent<TStepOutput extends Record<string, Json>>(
    id: string,
    spec: Omit<AgentStepSpec<TOutput, TStepOutput>, 'id' | 'type'>,
  ): StepsBuilder<TInput, TStepOutput> {
    return this.step({
      ...spec,
      id,
      type: 'agent',
    })
  }

  public branch<TStepOutput extends Record<string, Json>>(
    id: string,
    config: Omit<BranchStepSpec<TOutput, TStepOutput>, 'id' | 'type'>,
  ): StepsBuilder<TInput, TStepOutput> {
    return this.step({
      ...config,
      id,
      type: 'branch',
    })
  }
}

export const builder = <TInput extends Record<string, Json>>(): StepsBuilder<TInput, TInput> => {
  return new StepsBuilder()
}
