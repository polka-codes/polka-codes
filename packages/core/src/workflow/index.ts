// generated by polka.codes

export * from './agent'
export * from './builder'
export * from './command'
export * from './handlers'
export * from './dsl'
export * from './steps'
export * from './types'
export * from './workflow'
