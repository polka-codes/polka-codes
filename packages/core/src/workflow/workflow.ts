// generated by polka.codes

import { runStep } from './runStep'
import type { Json, StepSpecHandlerFn, WorkflowContext, WorkflowRunResult, WorkflowSpec } from './types'

export const run = async <TInput extends Record<string, Json>, TOutput extends Record<string, Json>>(
  workflow: WorkflowSpec<TInput, TOutput>,
  context: WorkflowContext,
  handler: StepSpecHandlerFn<any, any>,
  input: TInput,
): Promise<WorkflowRunResult<TOutput>> => {
  const rootStep = handler(workflow.step, handler)
  const result = await runStep(rootStep, input, context, undefined, {})
  switch (result.type) {
    case 'paused':
      return {
        type: 'paused',
        state: result.state,
      }
    case 'error':
      return {
        type: 'error',
        error: result.error,
      }
    case 'success':
      return {
        type: 'success',
        output: result.output as TOutput,
      }
  }
}

export const resume = async <TInput extends Record<string, Json>, TOutput extends Record<string, Json>>(
  workflow: WorkflowSpec<TInput, TOutput>,
  context: WorkflowContext,
  handler: StepSpecHandlerFn<any, any>,
  state: any,
  input: TInput, // This is the new input for the resumed workflow
): Promise<WorkflowRunResult<TOutput>> => {
  const rootStep = handler(workflow.step, handler)
  const result = await runStep(rootStep, input, context, state, {})
  switch (result.type) {
    case 'paused':
      return {
        type: 'paused',
        state: result.state,
      }
    case 'error':
      return {
        type: 'error',
        error: result.error,
      }
    case 'success':
      return {
        type: 'success',
        output: result.output as TOutput,
      }
  }
}
