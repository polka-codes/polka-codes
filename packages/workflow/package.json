{
  "name": "@polka-codes/workflow",
  "version": "0.9.41",
  "license": "AGPL-3.0",
  "author": "github@polka.codes",
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    }
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "bun build src/index.ts --outdir dist --target node"
  },
  "dependencies": {
    "@polka-codes/core": "workspace:*"
  }
}
