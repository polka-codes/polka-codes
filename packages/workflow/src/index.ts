// generated by polka.codes

export * from './agent.workflow'
export * from './helpers'
export * from './json-ai-types'
export * from './v2'
export * from './workflow'
