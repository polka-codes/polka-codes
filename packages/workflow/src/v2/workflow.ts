import type { Logger, ToolRegistry } from '../workflow'

export type ToolHandler<TTools extends ToolRegistry> = {
  [K in keyof TTools]: (input: TTools[K]['input']) => Promise<TTools[K]['output']>
}

export type WorkflowContextV2<TTools extends ToolRegistry> = {
  step: <T>(name: string, fn: () => Promise<T>) => Promise<T>
  logger: Logger
  toolHandler: ToolHandler<TTools>
}

export type WorkflowFnV2<TInput, TOutput, TTools extends ToolRegistry> = (
  input: TInput,
  context: WorkflowContextV2<TTools>,
) => Promise<TOutput>

// Create a default silent logger
const silentLogger: Logger = {
  debug: () => {},
  info: () => {},
  warn: () => {},
  error: () => {},
}

export function createContext<TTools extends ToolRegistry>(
  toolHandler: ToolHandler<TTools>,
  stepFn?: <T>(name: string, fn: () => Promise<T>) => Promise<T>,
  logger: Logger = silentLogger,
): WorkflowContextV2<TTools> {
  if (!stepFn) {
    // simple default step function
    stepFn = async <T>(_name: string, fn: () => Promise<T>) => fn()
  }

  return { step: stepFn, logger, toolHandler }
}

export const makeStepFn = (): (<T>(name: string, fn: () => Promise<T>) => Promise<T>) => {
  const results: Map<string, any> = new Map()

  return async <T>(name: string, fn: () => Promise<T>): Promise<T> => {
    if (results.has(name)) {
      return results.get(name) as T
    }

    const result = await fn()
    results.set(name, result)
    return result
  }
}
