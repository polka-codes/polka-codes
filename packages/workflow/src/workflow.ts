type ValidJsonOrVoid<T> = T extends string | number | boolean | null | void
  ? T
  : T extends readonly (infer U)[]
    ? readonly ValidJsonOrVoid<U>[]
    : // biome-ignore lint/complexity/noBannedTypes: this is needed
      T extends Function | symbol | bigint | Date | Map<any, any> | Set<any> | WeakMap<any, any> | WeakSet<any>
      ? never
      : T extends object
        ? { readonly [K in keyof T]: ValidJsonOrVoid<T[K]> }
        : never

export type StepFn<TTools extends ToolRegistry = ToolRegistry> = <T>(
  name: string,
  fn: () => AsyncGenerator<ToolCall<TTools>, ValidJsonOrVoid<T>, TTools[keyof TTools]['output']>,
) => AsyncGenerator<ToolCall<TTools>, T, TTools[keyof TTools]['output']>

export type ToolSignature<I, O> = {
  input: I
  output: O
}

export type ToolRegistry = Record<string, ToolSignature<any, any>>

export type ToolCall<TTools extends ToolRegistry> = {
  [K in keyof TTools]: {
    type: 'tool'
    tool: K
    input: TTools[K]['input']
  }
}[keyof TTools]

export type ToolsExecutor<TTools extends ToolRegistry> = {
  [K in keyof TTools]: (input: TTools[K]['input']) => Promise<TTools[K]['output']>
}

export type WorkflowTools<TTools extends ToolRegistry> = {
  [K in keyof TTools]: (input: TTools[K]['input']) => Generator<
    {
      type: 'tool'
      tool: K
      input: TTools[K]['input']
    },
    TTools[K]['output'],
    TTools[K]['output']
  >
}

export interface Logger {
  debug: (...args: any[]) => void
  info: (...args: any[]) => void
  warn: (...args: any[]) => void
  error: (...args: any[]) => void
}

export type WorkflowContext<TTools extends ToolRegistry> = {
  step: StepFn<TTools>
  tools: WorkflowTools<TTools>
  logger: Logger
}

export type WorkflowFn<TInput, TOutput, TTools extends ToolRegistry> = (
  input: TInput,
  context: WorkflowContext<TTools>,
) => AsyncGenerator<ToolCall<TTools>, TOutput, TTools[keyof TTools]['output']>

export type Workflow<TInput, TOutput, TTools extends ToolRegistry> = {
  name: string
  description: string
  fn: WorkflowFn<TInput, TOutput, TTools>
}

export type WorkflowStatusPending<TTools extends ToolRegistry> = {
  status: 'pending'
  tool: ToolCall<TTools>
}

export type WorkflowStatusCompleted<TOutput> = {
  status: 'completed'
  output: TOutput
}

export type WorkflowStatusFailed = {
  status: 'failed'
  error: any
}

export type WorkflowStatus<TTools extends ToolRegistry, TOutput> =
  | WorkflowStatusPending<TTools>
  | WorkflowStatusCompleted<TOutput>
  | WorkflowStatusFailed

import { makeStepFn } from './helpers'

export type WorkflowResult<TInput, TOutput, TTools extends ToolRegistry> =
  | (WorkflowStatusPending<TTools> & {
      next: (toolResult: TTools[keyof TTools]['output']) => Promise<WorkflowResult<TInput, TOutput, TTools>>
      throw: (error: Error) => Promise<WorkflowResult<TInput, TOutput, TTools>>
    })
  | WorkflowStatusCompleted<TOutput>
  | WorkflowStatusFailed

// Create a default silent logger
const silentLogger: Logger = {
  debug: () => {},
  info: () => {},
  warn: () => {},
  error: () => {},
}

export async function run<TInput, TOutput, TTools extends ToolRegistry>(
  workflow: Workflow<TInput, TOutput, TTools>,
  input: TInput,
  stepFn?: StepFn<TTools>,
  logger: Logger = silentLogger,
): Promise<WorkflowResult<TInput, TOutput, TTools>> {
  if (!stepFn) {
    stepFn = makeStepFn<TTools>()
  }

  const tools = new Proxy({} as WorkflowTools<TTools>, {
    get: (_target, tool: string) => {
      return function* (input: any): Generator<ToolCall<TTools>, any, any> {
        return yield {
          type: 'tool',
          tool,
          input,
        }
      }
    },
  })

  const context: WorkflowContext<TTools> = { step: stepFn, tools, logger }
  const gen = workflow.fn(input, context)

  let status: WorkflowStatus<TTools, TOutput>
  try {
    const { value, done } = await gen.next()
    if (done) {
      const status: WorkflowStatusCompleted<TOutput> = { status: 'completed', output: value }
      return status
    }
    status = { status: 'pending', tool: value }
  } catch (e) {
    status = { status: 'failed', error: e }
    return status
  }

  const next: (toolResult: TTools[keyof TTools]['output']) => Promise<WorkflowResult<TInput, TOutput, TTools>> = async (
    toolResult: TTools[keyof TTools]['output'],
  ) => {
    switch (status.status) {
      case 'pending': {
        try {
          const { value, done } = await gen.next(toolResult)
          if (done) {
            status = { status: 'completed', output: value }
            return status
          }
          status = { status: 'pending', tool: value }
        } catch (e) {
          status = { status: 'failed', error: e }
          return status
        }
        return { ...status, next, throw: throwError }
      }
      case 'completed':
      case 'failed':
        return status
    }
  }

  const throwError: (error: Error) => Promise<WorkflowResult<TInput, TOutput, TTools>> = async (error: Error) => {
    switch (status.status) {
      case 'pending': {
        try {
          const { value, done } = await gen.throw(error)
          if (done) {
            status = { status: 'completed', output: value }
            return status
          }
          status = { status: 'pending', tool: value }
        } catch (e) {
          status = { status: 'failed', error: e }
          return status
        }
        return { ...status, next, throw: throwError }
      }
      case 'completed':
      case 'failed':
        return status
    }
  }

  if (status.status === 'pending') {
    return {
      ...status,
      next,
      throw: throwError,
    }
  }
  return status
}
