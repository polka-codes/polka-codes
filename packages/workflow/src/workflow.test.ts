// generated by polka.codes
import { expect, spyOn, test } from 'bun:test'
import { createContext, makeStepFn, type ToolHandler, type ToolSignature, type WorkflowFn } from './workflow'

type Tools = {
  add: ToolSignature<{ a: number; b: number }, number>
  concat: ToolSignature<{ a: string; b: string }, string>
}

const simpleWorkflow: WorkflowFn<string, string, Tools> = async function simple(input, { step: _step }) {
  return `hello ${input}`
}

const singleToolWorkflow: WorkflowFn<number, number, Tools> = async function singleTool(input, { step: _step, toolHandler }) {
  const result = await toolHandler.add({ a: input, b: 2 })
  return result
}

const multiToolWorkflow: WorkflowFn<string, string, Tools> = async function multiTool(input, { step: _step, toolHandler }) {
  const r1 = await toolHandler.concat({ a: 'hello', b: ' ' })
  const r2 = await toolHandler.concat({ a: r1, b: input })
  return r2
}

const failingWorkflow: WorkflowFn<null, null, Tools> = async function failing() {
  throw new Error('test error')
}

const failingAfterToolWorkflow: WorkflowFn<number, number, Tools> = async function failingAfterTool(input, { step: _step, toolHandler }) {
  await toolHandler.add({ a: input, b: 2 })
  throw new Error('failed after tool')
}

const mockToolHandler: ToolHandler<Tools> = {
  add: async (input) => {
    return input.a + input.b
  },
  concat: async (input) => {
    return input.a + input.b
  },
}

test('should run a simple workflow', async () => {
  const context = createContext(mockToolHandler)
  const result = await simpleWorkflow('world', context)
  expect(result).toBe('hello world')
})

test('should run a workflow with one tool', async () => {
  const context = createContext(mockToolHandler)
  const result = await singleToolWorkflow(3, context)
  expect(result).toBe(5)
})

test('should run a workflow with multiple tools', async () => {
  const context = createContext(mockToolHandler)
  const result = await multiToolWorkflow('world', context)
  expect(result).toBe('hello world')
})

test('should handle a failing workflow', async () => {
  const context = createContext(mockToolHandler)
  await expect(failingWorkflow(null, context)).rejects.toThrow('test error')
})

test('should handle failing after a tool call', async () => {
  const context = createContext(mockToolHandler)
  await expect(failingAfterToolWorkflow(3, context)).rejects.toThrow('failed after tool')
})

test('should execute step for each call', async () => {
  const mock = {
    innerFn: async () => 'result',
  }
  const innerFnSpy = spyOn(mock, 'innerFn')

  const stepWorkflow: WorkflowFn<null, string, Tools> = async function stepWorkflow(_input, { step }) {
    const r1 = await step('step1', async () => {
      return await mock.innerFn()
    })
    const r2 = await step('step2', async () => {
      return await mock.innerFn()
    })
    return `${r1}-${r2}`
  }

  const context = createContext(mockToolHandler)
  const result = await stepWorkflow(null, context)
  expect(result).toBe('result-result')
  expect(innerFnSpy).toHaveBeenCalledTimes(2)
})

test('should support steps in a loop', async () => {
  const mock = {
    innerFn: async (i: number) => `result-${i}`,
  }
  const innerFnSpy = spyOn(mock, 'innerFn')

  const loopWorkflow: WorkflowFn<number, string, Tools> = async function loopWorkflow(input, { step }) {
    const results: string[] = []
    for (let i = 0; i < input; i++) {
      const r = await step(`loopedStep-${i}`, async () => {
        return await mock.innerFn(i)
      })
      results.push(r)
    }
    return results.join(',')
  }

  const context = createContext(mockToolHandler)
  const result = await loopWorkflow(3, context)
  expect(result).toBe('result-0,result-1,result-2')
  expect(innerFnSpy).toHaveBeenCalledTimes(3)
})

test('makeStepFn should cache results for same step name', async () => {
  const mock = {
    innerFn: async () => 'result',
  }
  const innerFnSpy = spyOn(mock, 'innerFn')
  const step = makeStepFn()

  const r1 = await step('step1', mock.innerFn)
  expect(r1).toBe('result')
  expect(innerFnSpy).toHaveBeenCalledTimes(1)

  // Call again with same step name
  const r2 = await step('step1', mock.innerFn)
  expect(r2).toBe('result')
  expect(innerFnSpy).toHaveBeenCalledTimes(1) // Should not be called again

  // Call with a different step name
  const r3 = await step('step2', mock.innerFn)
  expect(r3).toBe('result')
  expect(innerFnSpy).toHaveBeenCalledTimes(2) // Should be called again
})

test('makeStepFn should handle nested calls and cache them correctly', async () => {
  const mock = {
    innerFn: async () => 'inner-result',
    outerFn: async (s: string) => `outer-result-with-${s}`,
  }
  const innerFnSpy = spyOn(mock, 'innerFn')
  const outerFnSpy = spyOn(mock, 'outerFn')

  const step = makeStepFn()

  // First call to outer, which calls inner
  const r1 = await step('outer', async () => {
    const innerResult = await step('inner', mock.innerFn)
    return mock.outerFn(innerResult)
  })

  expect(r1).toBe('outer-result-with-inner-result')
  expect(innerFnSpy).toHaveBeenCalledTimes(1)
  expect(outerFnSpy).toHaveBeenCalledTimes(1)

  // Second call to outer, should be fully cached
  const r2 = await step('outer', async () => {
    const innerResult = await step('inner', mock.innerFn)
    return mock.outerFn(innerResult)
  })
  expect(r2).toBe('outer-result-with-inner-result')
  expect(innerFnSpy).toHaveBeenCalledTimes(1)
  expect(outerFnSpy).toHaveBeenCalledTimes(1)

  // Call inner again, should NOT be cached as it has a different call stack
  const r3 = await step('inner', mock.innerFn)
  expect(r3).toBe('inner-result')
  expect(innerFnSpy).toHaveBeenCalledTimes(2)
})

test('makeStepFn should not cache errors', async () => {
  let shouldThrow = true
  const mock = {
    innerFn: async () => {
      if (shouldThrow) {
        throw new Error('test error')
      }
      return 'result'
    },
  }
  const innerFnSpy = spyOn(mock, 'innerFn')

  const step = makeStepFn()

  // First call, should throw
  await expect(step('step1', mock.innerFn)).rejects.toThrow('test error')
  expect(innerFnSpy).toHaveBeenCalledTimes(1)

  // Second call, should not throw and should execute again
  shouldThrow = false
  const r2 = await step('step1', mock.innerFn)
  expect(r2).toBe('result')
  expect(innerFnSpy).toHaveBeenCalledTimes(2)
})
