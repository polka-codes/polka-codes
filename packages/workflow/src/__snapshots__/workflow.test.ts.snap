// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`should run a simple workflow 1`] = `
{
  "output": "hello world",
  "status": "completed",
}
`;

exports[`should run a workflow with one tool 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "a": 3,
      "b": 2,
    },
    "tool": "add",
    "type": "tool",
  },
}
`;

exports[`should run a workflow with one tool 2`] = `
{
  "output": 5,
  "status": "completed",
}
`;

exports[`should run a workflow with multiple tools 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "a": "hello",
      "b": " ",
    },
    "tool": "concat",
    "type": "tool",
  },
}
`;

exports[`should run a workflow with multiple tools 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "a": "hello ",
      "b": "world",
    },
    "tool": "concat",
    "type": "tool",
  },
}
`;

exports[`should run a workflow with multiple tools 3`] = `
{
  "output": "hello world",
  "status": "completed",
}
`;

exports[`should handle a failing workflow 1`] = `
{
  "error": "test error",
  "status": "failed",
}
`;

exports[`should handle failing after a tool call 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "a": 3,
      "b": 2,
    },
    "tool": "add",
    "type": "tool",
  },
}
`;

exports[`should handle failing after a tool call 2`] = `
{
  "error": "failed after tool",
  "status": "failed",
}
`;

exports[`should support tool calls in step functions 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "a": 5,
      "b": 10,
    },
    "tool": "add",
    "type": "tool",
  },
}
`;

exports[`should support tool calls in step functions 2`] = `
{
  "output": 15,
  "status": "completed",
}
`;

exports[`should support multiple tool calls in step functions 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "a": 3,
      "b": 5,
    },
    "tool": "add",
    "type": "tool",
  },
}
`;

exports[`should support multiple tool calls in step functions 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "a": 8,
      "b": 10,
    },
    "tool": "add",
    "type": "tool",
  },
}
`;

exports[`should support multiple tool calls in step functions 3`] = `
{
  "output": 18,
  "status": "completed",
}
`;

exports[`should support mixing regular and generator steps 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "a": 5,
      "b": 100,
    },
    "tool": "add",
    "type": "tool",
  },
}
`;

exports[`should support mixing regular and generator steps 2`] = `
{
  "output": 210,
  "status": "completed",
}
`;

exports[`should handle errors in generator steps 1`] = `
{
  "error": "error in generator step",
  "status": "failed",
}
`;
